Add support for shared-object library builds

EricEngle-NOAA opened the pull request:
Changes to allow for shared-object library building.  Shared and static libraries can be built in the same build run.  Other changes were necessary for successful testing.

No changes to the GitHub Workflow CI config here because it depends on an updated NCEPLIBS-sp that supports shared objects, [described here](https://github.com/NOAA-EMC/NCEPLIBS-sp/issues/67).  Therefore GitHub CI tests could fail.

EricEngle-NOAA committed:
Update cmake config for tests to get sp

Added target_link_libraries command pointing to sp::sp_d wherever the
same command existed for ip::ip_d.

This fixes undefined reference issues with the test programs.

EricEngle-NOAA committed:
Update for top-level cmake config

Added support for option BUILD_SHARED_LIBS

EricEngle-NOAA committed:
Update for src cmake config file

This update modifies the build process to first build object libraries,
then build static and shared (if BUILD_SHARED_LIBS=ON) libraries from
the object library.

This allows for the building of static and shared object libraries in
the same build run.

EricEngle-NOAA committed:
Updates for ipolate[s,v].F90 and ipxetas.F90

ipolates.F90 and ipolatev.F90:

Added single field versions of the grib2 interpolator subroutines.
Single field versions existed for grib1, but not grib2.  These specific
subroutines have been added to the generic interfaces ipolates and
ipolatev.

ipxetas.F90 this calls ipolates for grib2 and only passes a single
field, so this now compiles with no error.  Also made IBI and IBO
scalars instead on rank-1 array with size 1.

edwardhartnett commented:
Thanks for your help with the ip library! Since Kyle left, it is sadly neglected...

Can you please add two issues to describe this work, one for the changes to make shared libraries, the other to describe the new code you are adding to the ip library?

EricEngle-NOAA commented:
Sorry about that.  I have created new issues, #129 and #130

edwardhartnett commented:
OK, you are getting this warning in the documentatioon build:

```
warning: Tag 'TCL_SUBST' at line 260 of file '/home/runner/work/NCEPLIBS-ip/NCEPLIBS-ip/ip/build/docs/Doxyfile' has become obsolete.
32
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
33
warning: Tag 'COLS_IN_ALPHA_INDEX' at line 1081 of file '/home/runner/work/NCEPLIBS-ip/NCEPLIBS-ip/ip/build/docs/Doxyfile' has become 

```

To fix, go to the docs directory and run:

doxygen -u Doxyfile.in

This will update the doxyfile and silence the warnings. Since we are checking the documentation for completeness, all documentation warnings are fatal.

edwardhartnett commented:
OK, there are a bunch of CI issues...

This is timely because I have been going around the NCEPLIBS and upgrading the CI systems to new standards. Stand by while I get it working in #132 ...

edwardhartnett commented:
Once #132 is reviewed and merged, you can merge it into this PR and try again, and it should go much better.

BTW please do not introduce any new Fortran warnings. I don't think we managed to get rid of existing warnings so we're not doing a warning check as I would like to...

EricEngle-NOAA committed:
Updating Doxyfile.in

edwardhartnett commented:
OK, checkout develop and merge it to this branch, and you'll get the new, functional CI. Throw away your Doxygile.in changes, that's already fixed in develop.

EricEngle-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

edwardhartnett commented:
`/home/runner/work/NCEPLIBS-ip/NCEPLIBS-ip/ip/src/ipolates.F90:636: error: The following parameters of ipolates_mod::ipolates_grib2_single_field(integer, intent(in) IP, integer, dimension(20), intent(in) IPOPT, integer, intent(in) IGDTNUMI, integer, dimension(igdtleni), intent(in) IGDTMPLI, integer, intent(in) IGDTLENI, integer, intent(in) IGDTNUMO, integer, dimension(igdtleno), intent(in) IGDTMPLO, integer, intent(in) IGDTLENO, integer, intent(in) MI, integer, intent(in) MO, integer, intent(in) KM, integer, intent(in) IBI, logical *1, dimension(mi), intent(in) LI, real, dimension(mi), intent(in) GI, integer, intent(out) NO, real, dimension(mo), intent(inout) RLAT, real, dimension(mo), intent(inout) RLON, integer, intent(out) IBO, logical *1, dimension(mo), intent(out) LO, real, dimension(mo), intent(out) GO, integer, intent(out) IRET) are not documented:`

edwardhartnett commented on docs/Doxyfile.in:
```diff
@@ -1099,6 +1099,44 @@ USE_HTAGS              = NO
 
 VERBATIM_HEADERS       = YES
```
Git rid of this file in your PR. The Doxyfile.in on develop now matched the doxygen version of ubuntu-latest on the GitHub CI actions. So you should not need to change it.

edwardhartnett commented on src/ipolates.F90:
```diff
@@ -590,5 +591,87 @@ SUBROUTINE IPOLATES_grib2(IP,IPOPT,IGDTNUMI,IGDTMPLI,IGDTLENI, &
+  !> Special case of ipolates_grib2 when interpolating a single field.
+  !! Removes the km dimension of input arrays so scalars can be passed to ibi/ibo.
+  !!
+  !! @param[in] igdtleni Number of elements of the grid definition
```
THe order and names of the @param items have to match the variables in the subroutine argument list.

EricEngle-NOAA committed:
Turning off VERBATIM_HEADERS.

EricEngle-NOAA committed:
Updating docblocks.

EricEngle-NOAA commented:
I think MacOS is failing because sp is being build without OpenMP, but it is turned on for ip.
https://github.com/NOAA-EMC/NCEPLIBS-ip/blob/953e2d9fb951d9b07a792e4822664364838d6299/.github/workflows/MacOS.yml#L44

EricEngle-NOAA commented:
Possibly the same with Intel.  I can take a shot at updating these.

EricEngle-NOAA committed:
Update for Intel and MacOS GitHub CI config.

edwardhartnett commented on src/ipolates.F90:
```diff
@@ -590,5 +591,215 @@ SUBROUTINE IPOLATES_grib2(IP,IPOPT,IGDTNUMI,IGDTMPLI,IGDTLENI, &
+  !! - 3x Invalid budget method parameters.
+  !! - 4x Invalid spectral method parameters.
+  !!
+  !! @author Mark Iredell, Kyle Gerheiser, Eric Engle
```
Since you are the one adding this function, only your name should appear as author. It's understood that the code derives from other existing code, so you don't have to explicitly credit Mark and Kyle as authors. Same for any other functions you are adding...

edwardhartnett commented on src/ipolatev.F90:
```diff
@@ -702,5 +703,130 @@ subroutine ipolatev_grib1_single_field(ip,ipopt,kgdsi,kgdso,mi,mo,km,ibi,li,ui,v
+  !> - 3x Invalid budget method parameters.
+  !> - 4x Invalid spectral method parameters.
+  !>
+  !> @author Kyle Gerheiser @date July 2021
```
Did Kyle write this or did you? Who is the author? Since you are adding the function, I believe it should be you, unless this is some code that Kyle wrote before he left and never checked in...

edwardhartnett commented on docs/Doxyfile.in:
```diff
@@ -1097,7 +1097,45 @@ USE_HTAGS              = NO
 # See also: Section \class.
 # The default value is: YES.
 
-VERBATIM_HEADERS       = YES
```
You should have no changes to the Doxyfile.in at all, I believe.

It should not appear in your PR. The way I would normally do that is to checkout develop branch in another directory, copy the Doxyfile.in into the directory with this branch, and then do a commit. The doxyfile.in on your branch will then be identical to the one in develop, and the Doxyfile.in will disappear from the PR.

edwardhartnett commented on .github/workflows/Intel.yml:
```diff
@@ -63,7 +63,7 @@ jobs:
         mkdir build
         cd build
-        cmake .. -DCMAKE_INSTALL_PREFIX=~/sp
+        cmake .. -DOPENMP=ON -DBUILD_SHARED_LIBS=ON -DCMAKE_INSTALL_PREFIX=~/sp
```
I think we need to test both with and without shared libs. Let's get your PR merged and then I will circle back around and change the CI system to build with both shared and not shared.

EricEngle-NOAA committed:
Update Doxyfile.in

EricEngle-NOAA committed:
Update ipolates.F90

EricEngle-NOAA committed:
Update ipolatev.F90

EricEngle-NOAA committed:
Update ipolates.F90

EricEngle-NOAA commented:
Thanks for being patient with these changes.

EricEngle-NOAA committed:
Update cmake config

Updated source cmake configuration to add target_link_libraries() pointing to sp for static and shared libs.

EricEngle-NOAA committed:
Merge branch 'develop' into develop

EricEngle-NOAA commented:
Looks like https://github.com/NOAA-EMC/NCEPLIBS-ip/pull/128/commits/c0b4acbe209b54041c1a7ebba377bc9d4edcf052 from earlier fixed the issues with building on macOS.

EricEngle-NOAA committed:
Update Intel.yml

EricEngle-NOAA committed:
Update Intel.yml

EricEngle-NOAA committed:
Update for test Cmake config

Added set_property for the 2 C executables to explicitly let CMake know
that these are C programs.

EricEngle-NOAA commented:
Finally got the Intel tests to pass!

EricEngle-NOAA commented on tests/CMakeLists.txt:
```diff
@@ -33,6 +33,7 @@ target_link_libraries(test_library_grib2 PUBLIC sp::sp_d)
 target_link_libraries(test_library_grib2 PUBLIC ip::ip_d)
 
 add_executable(test_gdswzd_grib2 test_gdswzd_grib2.c)
+set_property(TARGET test_gdswzd_grib2 PROPERTY LINKER_LANGUAGE C)
```
The errors kept telling me that it was trying to build a Fortran executable but clearly this C with int main() in the source.  So the solution is to explicitly tell CMake this is a C executable

edwardhartnett commented:
Is this PR ready to merge?

EricEngle-NOAA commented:
Yes this is ready to merge.

edwardhartnett commented:
OK there is a lot going on in this PR. ;-) I will try and get it merged once #137 has been merged and the CI system is working again.

In future may I suggest a series of smaller PRs, which are likely to be easier and get merged faster...

EricEngle-NOAA committed:
Merge branch 'develop' into develop

EricEngle-NOAA committed:
Update for static libs option

Updated top-level CMake config to add BUILD_STATIC_LIBS and src Cmake
config to build static libs when True; added SOVERSION for shared libs

Updated tests CMake config to add sp_d for target links.

EricEngle-NOAA commented:
With https://github.com/NOAA-EMC/NCEPLIBS-ip/pull/128/commits/818abe0a7f07553bd3e755d6dd9a2db7b86eb877, I also added BUILD_STATIC_LIBS option
